pkg/porter/archive.go

```diff
@@ -2,24 +2,172 @@ package porter
 
 import (
 	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
 
+	"github.com/deislabs/cnab-go/bundle"
+	"github.com/deislabs/cnab-go/imagestore"
+	"github.com/deislabs/cnab-go/imagestore/construction"
 	"github.com/deislabs/porter/pkg/context"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/pkg/errors"
 )
 
 // ArchiveOptions defines the valid options for performing an archive operation
 type ArchiveOptions struct {
-	bundleFileOptions
-	File string
+	BundleLifecycleOpts
+	ArchiveFile string
 }
 
 // Validate performs validation on the publish options
-func (o *ArchiveOptions) Validate(args []string, cxt *context.Context) error {
-	return nil
+func (o *ArchiveOptions) Validate(args []string, ctx *context.Context) error {
+	if len(args) < 1 || args[0] == "" {
+		return errors.New("Destination File is required")
+	}
+	o.ArchiveFile = args[0]
+	return o.BundleLifecycleOpts.Validate(args, ctx)
 }
 
 // Archive is a composite function that generates a CNAB thick bundle. It will pull the invocation image, and
 // any referenced images locally (if needed), export them to individual layers, generate a bundle.json and
 // then generate a gzipped tar archive containing the bundle.json and the images
 func (p *Porter) Archive(opts ArchiveOptions) error {
-	return fmt.Errorf("archive is not yet implemented")
+
+	err := p.prepullBundleByTag(&opts.BundleLifecycleOpts)
+	if err != nil {
+		return errors.Wrap(err, "unable to pull bundle before invoking credentials generate")
+	}
+
+	err = p.applyDefaultOptions(&opts.sharedOptions)
+	if err != nil {
+		return err
+	}
+	err = p.ensureLocalBundleIsUpToDate(opts.bundleFileOptions)
+	if err != nil {
+		return err
+	}
+
+	bun, err := p.CNAB.LoadBundle(opts.CNABFile, true)
+	if err != nil {
+		return errors.Wrap(err, "couldn't open bundle for archiving")
+	}
+	// This allows you to export thin or thick bundles, we only support generting "thick" archives
+	ctor, err := construction.NewConstructor(false)
+	if err != nil {
+		return err
+	}
+
+	dest, err := p.Config.FileSystem.OpenFile(opts.ArchiveFile, os.O_RDWR|os.O_CREATE, os.ModePerm)
+
+	exp := &exporter{
+		out:                   p.Config.Out,
+		logs:                  p.Config.Out,
+		bundle:                bun,
+		destination:           dest,
+		imageStoreConstructor: ctor,
+	}
+	if err := exp.export(); err != nil {
+		return err
+	}
+	// if ex.verbose {
```

Comment:
Should we just remove this section?

Reply:
i was going to add it back, eventually. just needs to be updated to work like we want it to work,